components/slack/slack.app.js

```diff
@@ -3,6 +3,243 @@ const { WebClient } = require('@slack/web-api')
+          }),
+          context: { types, cursor: resp.cursor, userNames },
+        }
+      },
+    },
+    user: {
+      type: "string",
+      label: "User",
+      //description: "Select one or more channels or DM conversations to monitor for new messages.",
+      async options({ prevContext }) {
+        let { types, cursor, userNames } = prevContext
+        if (types == null) {
+          const scopes = await this.scopes()
+          types = ["im"]
+          // TODO use paging
+          userNames = {}
+          for (const user of (await this.users()).users) {
+            userNames[user.id] = user.name
+          }
+        }
+        const resp = await this.availableConversations(types.join(), cursor)
+        return {
+          options: resp.conversations.map((c) => {
+            return { label: `@${userNames[c.user]}`, value: c.id }
+          }),
+          context: { types, cursor: resp.cursor, userNames },
+        }
+      },
+    },
+    group: {
+      type: "string",
+      label: "Group",
+      //description: "Select one or more channels or DM conversations to monitor for new messages.",
+      async options({ prevContext }) {
+        let { types, cursor, userNames } = prevContext
+        if (types == null) {
+          const scopes = await this.scopes()
+          types = ["mpim"]
+          // TODO use paging
+          userNames = {}
+          for (const user of (await this.users()).users) {
+            userNames[user.id] = user.name
+          }
+        }
+        const resp = await this.availableConversations(types.join(), cursor)
+        return {
+          options: resp.conversations.map((c) => {
+            return { label: c.purpose.value, value: c.id }
+          }),
+          context: { types, cursor: resp.cursor, userNames },
+        }
+      },
+    },
+    conversation: {
+      type: "string",
+      label: "Channel",
+      description: "Select a public or private channel, or a user or group to direct message (DM).",
+      async options({ prevContext }) {
+        let { types, cursor, userNames } = prevContext
+        if (types == null) {
+          const scopes = await this.scopes()
+          types = ["public_channel"]
+          if (scopes.includes("groups:read")) {
+            types.push("private_channel")
+          }
+          if (scopes.includes("mpim:read")) {
+            types.push("mpim")
+          }
+          if (scopes.includes("im:read")) {
+            types.push("im")
+            // TODO use paging
+            userNames = {}
+            for (const user of (await this.users()).users) {
+              userNames[user.id] = user.name
+            }
+          }
+        }
+        const resp = await this.availableConversations(types.join(), cursor)
+        return {
+          options: resp.conversations.map((c) => {
+            if (c.is_im) {
+              return { label: `Direct messaging with: @${userNames[c.user]}`, value: c.id }
```

Comment:
A lookup here may be unavoidable if c.name doesn't return us anything useful.  However I think doing an on demand users.info rather than paging through the entire list probably make more sense for most use cases.